MaxText/kernels/megablox/gmm.py

```diff
@@ -0,0 +1,819 @@
+
+    def _accum(is_last_k_tile):
+      if is_last_k_tile:
+        mask_k_rem_lhs = partial(mask_k_rem, dim=1, quantize=lhs_quantize_dtype is not None)
+        mask_k_rem_rhs = partial(mask_k_rem, dim=int(transpose_rhs), quantize=rhs_quantize_dtype is not None)
+      else:
+        mask_k_rem_lhs = lambda x: x
+        mask_k_rem_rhs = lambda x: x
+
+      if isinstance(lhs, QTensor):
+        # loaded_lhs = aqt_pl.load_qtensor(lhs)
+        # Let qx: QTensor, qx = quant(x, 8 , ...)
+        # qx.dequant() == qx.qvalue * qx.scale ~= x
+        # Thus, setting qvalue to zero is equivalent to setting original tensor
+        # to zero.
+        qvalue = mask_k_rem_lhs(lhs.qvalue[...])
+        loaded_lhs = dataclasses.replace(lhs, qvalue=qvalue)
+        loaded_lhs = aqt_pl.load_qtensor(loaded_lhs)
+      else:
+        loaded_lhs = mask_k_rem_lhs(lhs[...]).astype(input_dtype)
+
+      if isinstance(rhs, QTensor):
+        qvalue = mask_k_rem_rhs(rhs.qvalue[...])
+        loaded_rhs = dataclasses.replace(rhs, qvalue=qvalue)
+        loaded_rhs = aqt_pl.load_qtensor(loaded_rhs)
+      else:
+        loaded_rhs = mask_k_rem_rhs(rhs[...]).astype(input_dtype)
+
+      acc_scratch[...] += aqt_pl.dot_general(
+          loaded_lhs,
+          loaded_rhs,
+          preferred_element_type=jnp.float32,
+          dimension_numbers=dot_general_dims,
+      )
+      if is_last_k_tile:
+        _store_accum()
+
+    lax.cond(
+        k_i == tiles_k - 1,
+        partial(_accum, True),
+        partial(_accum, False),
+    )
+
+  def lhs_transform_indices(n_i, grid_id, k_i, group_metadata, group_offset):
+    # lhs is (m, k). Load the [tm, tk] matrix for this m-tile.
+    group_offsets, group_ids, m_tile_ids = group_metadata
+    del n_i, group_offsets, group_ids, group_offset
+    return m_tile_ids[grid_id], k_i
+
+  def rhs_transform_indices(n_i, grid_id, k_i, group_metadata, group_offset):
+    # rhs is (num_groups, k, n). Load the [tk, tn] matrix based on the group id
+    # for this m-tile.
+    group_offsets, group_ids, m_tile_ids = group_metadata
+    del group_offsets, m_tile_ids
+    if transpose_rhs:
+      k_i, n_i = n_i, k_i
+
+    # NOTE: If we're working on only a shard of the rhs we need to adjust the
+    # group index we load from to account for this. The group_ids are in the
+    # "unsharded" domain.
+    return group_ids[grid_id] - group_offset[0], k_i, n_i
+
+  def out_transform_indices(n_i, grid_id, k_i, group_metadata, group_offset):
+    # out is (m, n). Load the [tm, tn] matrix for this m-tile.
+    group_offsets, group_ids, m_tile_ids = group_metadata
+    del k_i, group_offsets, group_ids, group_offset
+    return m_tile_ids[grid_id], n_i
+
+  out_block_spec = pl.BlockSpec((tm, tn), out_transform_indices)
+  if existing_out is None:
+    in_out_block_spec: Any = None
+    input_output_aliases = {}
+  else:
+    in_out_block_spec = out_block_spec
+    num_inputs = 6
```

Comment:
What's this hardcode 6 here?

Reply:
In Pallas, you can write back to the given input kernel using its reference.
In order to do so, you need to pass the mapping between the input and output so that you can write back.
If given `existing_out` becomes a output reference to write back, and the argument index of `existing_out` is six to the kernel. (Note that, group_metadata contains three tensor internally).

The  `6` was hardcoded in the original kernel implementation as well.

Reply:
I see. thanks for the explanation! Let's add `The argument index of existing_out is six to the kernel. Note that, group_metadata contains three tensor internally which makes its index six.` as a comment there?